### PR TITLE
Refactor pkg/generate

### PR DESCRIPTION
**Pull request by fanzhangio**
This PR is the first step of crd generation refactor focusing on moving `rbac`, `webhook` from `pkg/generate` to `pkg/` to align with `crd` in the same package level.

**Move below interface in separate PR**
Add a generic `Generator` interface in `pkg/internal/codegen` which is on top of `gengo` for all generation usages in controller-tool

### CODE REVIEW

**Comment by mengqiy** [comment]
There seems to be a problem with travis CI...
Let's try close and reopen.

**Comment by droot** [comment]
@fanzhangio I am having trouble reviewing the PR because it is very large PR. would have been nice to split it into multiple commits. Is it possible to do so ?

**Comment by droot** [comment]
@fanzhangio Are you planning to split it or you want me to review as-is ?

**Comment by droot** [comment]
@fanzhangio lets chat offline how do we want to move forward with this PR.

**Comment by fanzhangio** [comment]
@droot Sure, I marked it with `WIP` for better discussion.

File: cmd/controller-gen/main.go [hunk 1]
```diff
@@ -22,7 +22,7 @@ import (
 	"path/filepath"
 
 	"github.com/spf13/cobra"
-	crdgenerator "sigs.k8s.io/controller-tools/pkg/crd/generator"
+	"sigs.k8s.io/controller-tools/pkg/generate/crd"
```
Comment (droot) [inline cmd/controller-gen/main.go]: `generate` is not a good pkg name as per Go guidelines. Given that we have moved scaffolding pkgs to `kubebuilder` repo, it will be safe to assume that the pkgs that live here are related to generation. So may be move the others `rbac` and `webhook` under `pkg` ?

File: cmd/controller-gen/main.go [hunk 2]
```diff
@@ -84,7 +84,7 @@ Usage:
 }
 
 func newCRDCmd() *cobra.Command {
-	g := &crdgenerator.Generator{}
+	o := &crd.Options{}
```
Comment (droot) [inline cmd/controller-gen/main.go]: If the pkg name becomes `crd` then `Generator` seems like more appropriate to capture generator options ?

File: pkg/generate/crd/crd.go [hunk 3]
```diff
@@ -26,58 +26,50 @@ import (
...
-	crdutil "sigs.k8s.io/controller-tools/pkg/crd/util"
 	"sigs.k8s.io/controller-tools/pkg/internal/codegen"
 	"sigs.k8s.io/controller-tools/pkg/internal/codegen/parse"
-	"sigs.k8s.io/controller-tools/pkg/util"
+	util "sigs.k8s.io/controller-tools/pkg/util"
 )
 
-// Generator generates CRD manifests from API resource definitions defined in Go source files.
-type Generator struct {
+// Options receive optional arguments from command line
```
Comment (droot) [inline pkg/generate/crd/crd.go]: Sorry, but the new comment doesn't give much useful information.

File: pkg/generate/crd/crd.go [hunk 4]
```diff
@@ -26,58 +26,50 @@ import (
...
+type Options struct {
 	RootPath          string
 	OutputDir         string
 	Domain            string
 	Namespace         string
 	SkipMapValidation bool
-
+	apisPkg           string
```
Comment (droot) [inline pkg/generate/crd/crd.go]: The previous comment was very useful for code understanding. Is there a reason we are deleting that ?

File: pkg/generate/crd/crd.go [hunk 5]
```diff
@@ -86,70 +78,57 @@ func (c *Generator) ValidateAndInitFields() error {
...
 
 	return nil
 }
 
-// Do manages CRD generation.
-func (c *Generator) Do() error {
-	arguments := args.Default()
-	b, err := arguments.NewBuilder()
+// Run generates CRD and stores CRD into output dir.
+func (c *Options) Run() error {
```
Comment (droot) [inline pkg/generate/crd/crd.go]: I found the previous abstractions easier to grok. 
Generator takes input options to generate CRD vs Options having Run method to generate the CRD. I find previous one more intuitive.

File: cmd/controller-gen/main.go (hunk 1 above)
Comment (fanzhangio) [reply cmd/controller-gen/main.go]: Agree. So, I will move all other packages, like `rbac`, `webhook` into `pkg` in this PR.

File: cmd/controller-gen/main.go (hunk 2 above)
Comment (fanzhangio) [reply cmd/controller-gen/main.go]: Make sense. So we can keep generator approach, but just split the generation procedure into some exposed APIs. I also wanted to have a more generic `Generator` upfront, which wraps `gengo` for all kinds of generation in this repo. This kind of `Generator` can live in `pkg/internal/codegen`

File: pkg/generate/crd/crd.go (hunk 3 above)
Comment (fanzhangio) [reply pkg/generate/crd/crd.go]: Make sense.

File: pkg/generate/crd/crd.go (hunk 4 above)
Comment (fanzhangio) [reply pkg/generate/crd/crd.go]: OK. I will keep the comments.

File: pkg/generate/crd/crd.go (hunk 5 above)
Comment (fanzhangio) [reply pkg/generate/crd/crd.go]: I see. I will rebase and update the PR. How about we do it by a series of changes. First, I just focus on moving packages in this PR, and then refactor the `Generator` ?

File: pkg/internal/codegen/generator.go [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+}
+
+// NameSystems returns the name system used by the generators in this package.
+// e.g. black-magic
+func NameSystems() namer.NameSystems {
+	return namer.NameSystems{
+		"public": namer.NewPublicNamer(1),
+		"raw":    namer.NewRawNamer("", nil),
+	}
+}
```
Comment (droot) [inline pkg/internal/codegen/generator.go]: Can we move this in a separate PR ? preferrably with use-case in existing code to demonstrate the utility ?

Comment (fanzhangio) [reply pkg/internal/codegen/generator.go]: Sure. Thanks

**Comment by mengqiy** [comment]
Looks like the comments have been resolved.
/lgtm
